pkg/kubelet/volumemanager/metrics/metrics.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/kubernetes/pkg/kubelet/volumemanager/cache"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const pluginNameNotAvailable = "N/A"
+
+var (
+	registerMetrics sync.Once
+
+	totalVolumesDesc = prometheus.NewDesc(
+		"volume_manager_total_volumes",
+		"Number of volumes in Volume Manager",
+		[]string{"plugin_name", "state"},
+		nil,
+	)
+
+	reconstructVolumeErrorsMetric = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "volume_manager_reconstruct_volume_errors_total",
+			Help: "Amount of errors on ReconstructVolumeOperation",
+		},
+		[]string{"plugin_name"},
+	)
+)
+
+// Register registers Volume Manager metrics.
+func Register(asw cache.ActualStateOfWorld, dsw cache.DesiredStateOfWorld, pluginMgr *volume.VolumePluginMgr) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(reconstructVolumeErrorsMetric)
+		prometheus.MustRegister(&totalVolumesCollector{asw, dsw, pluginMgr})
+	})
+}
+
+type totalVolumesCollector struct {
+	asw       cache.ActualStateOfWorld
+	dsw       cache.DesiredStateOfWorld
+	pluginMgr *volume.VolumePluginMgr
+}
+
+var _ prometheus.Collector = &totalVolumesCollector{}
+
+// Describe implements the prometheus.Collector interface.
+func (c *totalVolumesCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- totalVolumesDesc
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *totalVolumesCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
How frequently does Prometheus call `Collect(...)`? If it is too frequent it could cause perf issues.

Reply:
The standard interval is 1m [1].

[1] https://prometheus.io/docs/prometheus/latest/configuration/configuration/#configuration-file